File: backend/models/achievement.go
package models

import (
	"encoding/json"
	"time"
)

type Achievement struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Icon        string          `json:"icon"`
	Criteria    json.RawMessage `json:"criteria"`
	SortOrder   int             `json:"sort_order"`
}

type UserAchievement struct {
	ID            string    `json:"id"`
	UserID        string    `json:"user_id"`
	AchievementID string    `json:"achievement_id"`
	EarnedAt      time.Time `json:"earned_at"`
}

type UserAchievementProgress struct {
	UserID        string          `json:"user_id"`
	AchievementID string          `json:"achievement_id"`
	CurrentValue  int             `json:"current_value"`
	TargetValue   int             `json:"target_value"`
	Metadata      json.RawMessage `json:"metadata"`
	UpdatedAt     time.Time       `json:"updated_at"`
}

type AchievementItem struct {
	ID          string     `json:"id"`
	Name        string     `json:"name"`
	Description string     `json:"description"`
	Icon        string     `json:"icon"`
	EarnedAt    *time.Time `json:"earned_at,omitempty"`
}

type ProgressItem struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	Description  string `json:"description"`
	Icon         string `json:"icon"`
	CurrentValue int    `json:"current_value"`
	TargetValue  int    `json:"target_value"`
}

type AchievementsResponse struct {
	Earned      []AchievementItem `json:"earned"`
	InProgress  []ProgressItem    `json:"in_progress"`
	NewlyEarned []string          `json:"newly_earned"`
}

type AchievementEvent struct {
	UserID    string
	HabitID   string
	Timestamp time.Time
}

File: mobile/lib/features/achievements/models/achievement.dart
class AchievementItem {
  final String id;
  final String name;
  final String description;
  final String icon;
  final DateTime? earnedAt;

  const AchievementItem({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    this.earnedAt,
  });

  factory AchievementItem.fromJson(Map<String, dynamic> json) => AchievementItem(
        id: json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String,
        icon: json['icon'] as String,
        earnedAt: json['earned_at'] != null
            ? DateTime.parse(json['earned_at'] as String)
            : null,
      );
}

class ProgressItem {
  final String id;
  final String name;
  final String description;
  final String icon;
  final int currentValue;
  final int targetValue;

  const ProgressItem({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.currentValue,
    required this.targetValue,
  });

  bool get isNearCompletion => (targetValue - currentValue) <= 2;

  factory ProgressItem.fromJson(Map<String, dynamic> json) => ProgressItem(
        id: json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String,
        icon: json['icon'] as String,
        currentValue: json['current_value'] as int,
        targetValue: json['target_value'] as int,
      );
}

class AchievementsData {
  final List<AchievementItem> earned;
  final List<ProgressItem> inProgress;
  final List<String> newlyEarned;

  const AchievementsData({
    required this.earned,
    required this.inProgress,
    required this.newlyEarned,
  });

  List<ProgressItem> get nudges =>
      inProgress.where((p) => p.isNearCompletion).toList();

  factory AchievementsData.fromJson(Map<String, dynamic> json) => AchievementsData(
        earned: (json['earned'] as List)
            .map((e) => AchievementItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        inProgress: (json['in_progress'] as List)
            .map((e) => ProgressItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        newlyEarned: List<String>.from(json['newly_earned'] as List),
      );
}
